src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs

```diff
@@ -189,32 +280,61 @@ public void Dispose()
 
         internal AbstractProject GetProject(ProjectId id)
         {
-            AbstractProject project;
-            _projectMap.TryGetValue(id, out project);
-            return project;
+            lock (_gate)
+            {
+                AbstractProject project;
+                _projectMap.TryGetValue(id, out project);
+                return project;
+            }
         }
 
         /// <summary>
         /// Add a project to the workspace.
+        /// If invoked on the foreground thread, the add is executed right away.
+        /// Otherwise, the add is scheduled on foreground task scheduler.
```

Comment:
I have attempted to do the minimum work to 1. avoid switching back to the foreground thread when any free threaded method is invoked on the tracker from background and 2. enable multithreaded access to its data structures - I agree we could do lot more work asynchronously, but that could come in future.